changelog/dmd.deprecate-case-variables.dd

```diff
@@ -0,0 +1,44 @@
+Using run-time variables as switch cases has been deprecated
+
+It's a feature that's not commonly used, and it goes against normal expectations of a switch statement.
+When using run-time variables:
+
+- Case statements can contain duplicates
+- Case statements can be mutated through mutable aliasing of a `const ref`
```

Comment:
Explain why not just try to avoid aliasing (e.g. ban ref)

Reply:
The gist of it is that, currently, I see run-time case variables as a half-baked extension to the classic switch statement. We can commit going forward and fix the issues, but for the sake of simplicity, I propose we go back, since the future is not commonly used as far as I know.

Reply:
On the one hand it's simpler in a sense that you have less choice, but on the other hand you now actually have more rules - it's not just an (say) AssignExpression any-more, it's now AssignExpression (+ rules).

Although you're right in that it's not a super-common thing to do, the one time I have used such a pattern profitably is as an intermediary step between two styles: Allowing runtime constants is a useful bit of plasticity IMO. 

Consider what happens if you have a large switch statement, and it turns out you were too clever and the design actually doesn't fit in a C-style (const-only) switch statement. We'd be encouraging people to write in a C++ style like
```
const specialCase = calc();
if (x == specialCase) {

}
switch(x) {

}
```
rather than just writing
```
const specialCase = calc();
switch(x) {
   case specialCase:
      return 6.6260701;
}
```

I am partial to the argument about duplication particularly in the case of `final` switch. You can try to prove such properties hold, but not an easy thing to do in dmd...

Reply:
> but on the other hand you now actually have more rules - it's not just an (say) AssignExpression any-more, it's now AssignExpression (+ rules).

Currently, these case types are supported:

| Case variable           | int     | string   | other types    |
|-------------------------|---------|----------|----------------|
| enum                    | yes     | yes      | no             |
| const var               | yes     | no       | no             |
| run-time expr / mut var | no      | no       | no             |

It's definitely not 'any AssignExpression'. After deprecating case variables:

| Case variable           | int     | string   | other types    |
|-------------------------|---------|----------|----------------|
| enum                    | yes     | yes      | no             |
| run-time expr   | no      | no       | no             |